tempesta_fw/t/unit/test_http_parser.c

```diff
@@ -1498,6 +1498,63 @@ TEST(http_parser, if_none_match)
 #undef ETAG_3
 }
 
+TEST(http_parser, referer)
+{
+	TfwHttpHdrTbl *ht;
+	TfwStr h_referer;
+
+	/*
+	 * URI scheme 'http://' is skipped during
+	 * parsing of 'Referer' header.
+	 */
+	const char *s_referer1 =
+		"tempesta-tech.com:8080"
+		"/cgi-bin/show.pl?entry=tempesta";
+	const char *s_referer2 =
+		"/cgi-bin/show.pl?entry=tempesta";
+	const char *s_referer3 =
+		"[2001:0db8:11a3:09d7:1f34:8a2e:07a0:765d]"
+		":8080/cgi-bin/show.pl?entry=tempesta";
+
+	FOR_REQ("GET /foo HTTP/1.1\r\n"
+		"Referer: http://tempesta-tech.com:8080"
+		"/cgi-bin/show.pl?entry=tempesta\r\n"
+		"\r\n")
+	{
+		ht = req->h_tbl;
+		tfw_http_msg_clnthdr_val(&ht->tbl[TFW_HTTP_HDR_REFERER],
+					 TFW_HTTP_HDR_REFERER,
+					 &h_referer);
+		EXPECT_TRUE(tfw_str_eq_cstr(&h_referer, s_referer1,
+					    strlen(s_referer1), 0));
+	}
+
+	FOR_REQ("GET /foo HTTP/1.1\r\n"
+		"Referer: /cgi-bin/show.pl?entry=tempesta\r\n"
+		"\r\n")
+	{
+		ht = req->h_tbl;
+		tfw_http_msg_clnthdr_val(&ht->tbl[TFW_HTTP_HDR_REFERER],
+					 TFW_HTTP_HDR_REFERER,
+					 &h_referer);
+		EXPECT_TRUE(tfw_str_eq_cstr(&h_referer, s_referer2,
+					    strlen(s_referer2), 0));
+	}
+
+	FOR_REQ("GET /foo HTTP/1.1\r\n"
+		"Referer: http://[2001:0db8:11a3:09d7:1f34:8a2e:07a0:765d]:8080"
```

Comment:
Try to add spaces before and after referrer link. It seems that the header  is not parsed correctly if any.

Reply:
Comment is still not addressed, try the patch below. The test fails with it:
```
[  203.687589] tfw_test: TEST_RUN(http_parser, referer)
[  203.689449] tfw_test: === request: [GET /foo HTTP/1.1
[  203.693618] tfw_test: /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_http_parser.c:1518: error: 
[  203.696479] tfw_test: FAIL:
[  203.697361] tfw_test:   test__http_parser__referer()
[  203.699459] tfw_test:   can't parse request (code=-2):
```
The patch:
```diff
diff --git a/tempesta_fw/t/unit/test_http_parser.c b/tempesta_fw/t/unit/test_http_parser.c
index 17d1ce1..c3c27c4 100644
--- a/tempesta_fw/t/unit/test_http_parser.c
+++ b/tempesta_fw/t/unit/test_http_parser.c
@@ -1504,8 +1504,8 @@ TEST(http_parser, referer)
        TfwStr h_referer;
 
        const char *s_referer1 =
-               "http://tempesta-tech.com:8080"
-               "/cgi-bin/show.pl?entry=tempesta";
+               "  http://tempesta-tech.com:8080"
+               "/cgi-bin/show.pl?entry=tempesta  ";
        const char *s_referer2 =
                "/cgi-bin/show.pl?entry=tempesta";
        const char *s_referer3 =
@@ -1513,8 +1513,8 @@ TEST(http_parser, referer)
                ":8080/cgi-bin/show.pl?entry=tempesta";
 
        FOR_REQ("GET /foo HTTP/1.1\r\n"
-               "Referer: http://tempesta-tech.com:8080"
-               "/cgi-bin/show.pl?entry=tempesta\r\n"
+               "Referer:   http://tempesta-tech.com:8080"
+               "/cgi-bin/show.pl?entry=tempesta  \r\n"
                "\r\n")
        {
                ht = req->h_tbl;
```

Reply:
Corrected.